test/tools/d_do_test.d

```diff
@@ -962,11 +966,94 @@ int runBashTest(string input_dir, string test_name)
     }
     return process.wait();
 }
+
+int runDShellTest(string input_dir, string test_name, const ref EnvData envData,
+    string output_dir, string output_file)
+{
+    const testScriptDir = buildPath(dmdTestDir, input_dir);
+    const testScriptPath = buildPath(testScriptDir, test_name ~ ".d");
+    const dshellPath = buildPath(dmdTestDir, "tools", "dshell.d");
+    const dshellSource = readText(dshellPath);
+
+    writefln(" ... %s/%s.d", input_dir, test_name);
+    if (exists(output_file))
+        remove(output_file);
+
+    // create the "dshellglue" module to glue dshell and the 'test' together
+    const glueSourceDir = buildPath(output_dir, test_name ~ ".dshellglue");
+    mkdirRecurse(glueSourceDir);
+    {
+        auto glueSource = File(buildPath(glueSourceDir, "dshellglue.d"), "w");
+        glueSource.writeln(`module dshellglue;`);
+        glueSource.writefln(`enum TEST_DIR  = "%s";`, input_dir);
+        glueSource.writefln(`enum TEST_NAME = "%s";`, test_name);
+    }
+
+    const testScriptExe = buildPath(output_dir, test_name ~ envData.exe);
+    const output_file_temp = output_file ~ ".tmp";
+    mkdirsFor(output_file_temp);
+
+    bool picFlag = false;
+    version (Windows) { } else
+    {
+        version(X86_64)
+            picFlag = true;
+        if (environment.get("PIC", null) == "1")
+            picFlag = true;
+    }
+
+    int exitCode;
+    {
+        auto outfile = File(output_file_temp, "w");
+        const compile = [envData.dmd, "-conf=", "-m"~envData.model] ~
+            (picFlag ? ["-fPIC"] : []) ~ [
+            "-od" ~ envData.results_dir, "-of" ~ testScriptExe, "-i",
+            "-I=" ~ testScriptDir,
+            "-I=" ~ glueSourceDir,
+            testScriptPath, dshellPath];
+        outfile.writeln("[COMPILE_TEST] ", escapeShellCommand(compile));
+        // Note that spawnprocess closes the file, so it will need to be re-opened
+        // below when we run the test
+        auto compileProc = std.process.spawnProcess(compile, stdin, outfile, outfile);
+        exitCode = wait(compileProc);
+    }
+    if (exitCode == 0)
+    {
+        auto outfile = File(output_file_temp, "a");
+        const runTest = [testScriptExe];
+        outfile.writeln("[RUN_TEST] ", escapeShellCommand(runTest));
+        auto runTestProc = std.process.spawnProcess(runTest, stdin, outfile, outfile);
+        exitCode = wait(runTestProc);
+    }
+
+    if (exitCode == 0)
+        rename(output_file_temp, output_file);
+    else
+    {
+        writeln("==============================");
+        writefln("Test '%s/%s.d' failed. The logged output:", input_dir, test_name);
+        const output = readText(output_file_temp);
+        write(output);
+        if (!output.endsWith("\n"))
+              writeln();
+        writeln("==============================");
+        remove(output_file_temp);
+    }
+    return exitCode;
```

Comment:
This is quite a big function. I recommend splitting it up into multiple methods or a struct with multiple members for better readability.